selectbyregex.py

```diff
@@ -2,53 +2,44 @@
 import sublime
 import sublime_plugin
 
-def has_in_group(rx):
-	return '(!' in rx
-
-def has_sel(rx):
-	return '$_' in rx
-
-def unwrap_in(rx):
-	# Unwrap (!...) group, returns regex and flag if this group was found
-	return rx.replace('(!', '(?P<Select>')
 
 def unwrap_sel(rx, s):
 	# Unwrap $_ with regex, that match str
 	return rx.replace('$_', re.escape(s))
 
-def has_sel(rx):
-	return '$_' in rx
 
-def unwrap(rx, s):
-	# Unwrap both, 'unwrap_in' must be first
-	return unwrap_sel(unwrap_in(rx), s)
+class SelectByRegexBase(sublime_plugin.TextCommand):
 
-class SelectByRegexNext(sublime_plugin.TextCommand):
-	def run(self, edit, regex = None):
+	def run(self, edit, regex=None):
 		self.text = self.view.substr(sublime.Region(0, self.view.size()))
 		self.selections = [r for r in self.view.sel()]
 		self.view.sel().clear()
 		self.view.window().show_input_panel('regex', regex or '', self.on_done, self.on_change, self.on_cancel)
 
+	def mark_groups(self, match):
+		for number, group in list(enumerate(match.groups(), start=1)):
+			self.inner_regions.append(sublime.Region(match.start(number), match.end(number)))
+		self.outer_regions.append(sublime.Region(match.start(), match.end()))
+
+
+class SelectByRegexNext(SelectByRegexBase):
+
 	def on_done(self, rx):
 		self.view.erase_regions('select by regex, next, outer')
 		self.view.erase_regions('select by regex, next, inner')
 
 		self.view.sel().add_all(self.inner_regions)
 
 	def on_change(self, rx):
-		self.in_group = has_in_group(rx)
 		self.inner_regions = []
 		self.outer_regions = []
 		self.highlight_regions = []
 		for r in self.selections:
-			self.rx = re.compile(unwrap(rx, self.view.substr(r)), re.MULTILINE)
-			m = self.rx.search(self.text, min(r.a, r.b))
-			if m and r.a == r.b or m.end() <= max(r.a, r.b) or has_sel(rx): # Restrict to selection if it's not empty and doesn't have $_
-				if self.in_group:
-					if 'Select' in m.groupdict():
-						self.inner_regions.append(sublime.Region(m.start('Select'), m.end('Select')))
-						self.outer_regions.append(sublime.Region(m.start(), m.end()))
+			self.rx = re.compile(unwrap_sel(rx, self.view.substr(r)), re.MULTILINE)
+			m = self.rx.search(self.text, r.a)
+			if m and r.a == r.b or m.end() <= r.b:  # Restrict to selection if it's not empty
+				if len(m.groups()) > 1:
```

Comment:
Yes, you are right! And if to be honest may be `len(m.groups())` will be clearer.

Thanks!